packages/0xcert-ethereum-erc20-contracts/src/contracts/dapp-token.sol

```diff
@@ -0,0 +1,608 @@
+    supportedInterfaces[0x95d89b41] = true; // ERC20 symbol
+    supportedInterfaces[0x313ce567] = true; // ERC20 decimals
+  }
+
+  /**
+   * @dev Returns the name of the token.
+   */
+  function name()
+    external
+    view
+    returns (string memory _name)
+  {
+    _name = tokenName;
+  }
+
+  /**
+   * @dev Returns the symbol of the token.
+   */
+  function symbol()
+    external
+    view
+    returns (string memory _symbol)
+  {
+    _symbol = tokenSymbol;
+  }
+
+  /**
+   * @dev Returns the number of decimals the token uses.
+   */
+  function decimals()
+    external
+    view
+    returns (uint8 _decimals)
+  {
+    _decimals = tokenDecimals;
+  }
+
+  /**
+   * @dev Returns the total token supply.
+   */
+  function totalSupply()
+    external
+    view
+    returns (uint256 _totalSupply)
+  {
+    _totalSupply = tokenTotalSupply;
+  }
+
+  /**
+   * @dev Returns the account balance of another account with address _owner.
+   * @param _owner The address from which the balance will be retrieved.
+   */
+  function balanceOf(
+    address _owner
+  )
+    external
+    view
+    returns (uint256 _balance)
+  {
+    _balance = balances[_owner];
+  }
+
+  /**
+   * @dev Returns the amount which _spender is still allowed to withdraw from _owner.
+   * @param _owner The address of the account owning tokens.
+   * @param _spender The address of the account able to transfer the tokens.
+   */
+  function allowance(
+    address _owner,
+    address _spender
+  )
+    external
+    view
+    returns (uint256 _remaining)
+  {
+    _remaining = allowed[_owner][_spender];
+  }
+
+  /**
+   * @dev Sets whitelist state to an address
+   * @param _target Address we are setting the whitelist state.
+   * @param _state State we are setting. True means the address is whitelisted while false the
+   * oposite.
+   */
+  function setWhitelist(
+    address _target,
+    bool _state
+  )
+    external
+    hasAbilities(ABILITY_SET_WHITELISTED)
+  {
+    whitelistedRecipients[_target] = _state;
+    emit Whitelist(_target, _state);
+  }
+
+  /**
+   * @dev Sets migration caller state to an address
+   * @param _target Address we are setting the migration caller state.
+   * @param _state State we are setting. True means the address is a migration caller while false
+   * the oposite.
+   */
+  function seMigrationCaller(
+    address _target,
+    bool _state
+  )
+    external
+    hasAbilities(ABILITY_SET_MIGRATOR_ADDRESS)
+  {
+    migrationCallers[_target] = _state;
+    emit MigrationCaller(_target, _state);
+  }
+
+  /**
+   * @dev Sets address to which migration is done.
+   * @param _target Targeted address.
+   */
+  function setMigrateAddress(
+    address _target
+  )
+    external
+    hasAbilities(ABILITY_SET_MIGRATE_ADDRESS)
+  {
+    require(_target != address(0), NULL_ADDRESS);
+    migrateAddress = _target;
+  }
+
+  /**
+   * Migrate to a new Dapp token.
+   */
+  function migrate()
```

Comment:
This is a race condition, but we are the ones controlling it. Basically the only reason `migrateAddress` can be changed after being set from zero to an actual address is to have a backup. I don't believe this would ever happen but if it does it is only to mitigate damage at that point. So I think this is an acceptable scenario.

@xpepermint

Reply:
I think it's good. @fulldecent comments?